From: Liam Howlett <liam.howlett@oracle.com>
Subject: kernel/fork: detect mas_store() failure in dup_mmap()
Date: Fri, 22 Jul 2022 16:06:03 +0000

mas_store() should not fail, but protect against potential failures by
checking the maple state for mas_is_err().

Link: https://lkml.kernel.org/r/20220722160546.1478722-3-Liam.Howlett@oracle.com
Fixes: 57579b57de57 (kernel/fork: use maple tree for dup_mmap() during forking)
Signed-off-by: Liam R. Howlett <Liam.Howlett@oracle.com>
Cc: Yu Zhao <yuzhao@google.com>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 kernel/fork.c |    5 +++++
 1 file changed, 5 insertions(+)

--- a/kernel/fork.c~kernel-fork-use-maple-tree-for-dup_mmap-during-forking-fix
+++ a/kernel/fork.c
@@ -709,6 +709,8 @@ static __latent_entropy int dup_mmap(str
 		mas.index = tmp->vm_start;
 		mas.last = tmp->vm_end - 1;
 		mas_store(&mas, tmp);
+		if (mas_is_err(&mas))
+			goto fail_nomem_mas_store;
 
 		mm->map_count++;
 		if (!(tmp->vm_flags & VM_WIPEONFORK))
@@ -732,6 +734,9 @@ out:
 fail_uprobe_end:
 	uprobe_end_dup_mmap();
 	return retval;
+
+fail_nomem_mas_store:
+	unlink_anon_vmas(tmp);
 fail_nomem_anon_vma_fork:
 	mpol_put(vma_policy(tmp));
 fail_nomem_policy:
_
